File: 77.combinations.java
/*
 * @lc app=leetcode id=77 lang=java
 *
 * [77] Combinations
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> combine(int n, int k) {
        if (n <= 0 || n < k) return res;
        backtrack(1, n, k, new LinkedList<>());
        return res;
    }

    private void backtrack(int s, int n, int k, List<Integer> tmpList) {
        if (k == 0) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        for (int i = s; k <= n-i+1; ++i) {
            tmpList.add(i);
            backtrack(i+1, n, k-1, tmpList);
            tmpList.remove(tmpList.size()-1);
        }
    }
}
// @lc code=end

File: go/77.combinations.go
/*
 * @lc app=leetcode id=77 lang=golang
 *
 * [77] Combinations
 */

// @lc code=start
func combine(n int, k int) [][]int {
	res := [][]int{}
	var backtrack func(int, []int)
	backtrack = func(idx int, comb []int) {
		if len(comb) == k {
			tmp := make([]int, k)
			copy(tmp, comb)
			res = append(res, tmp)
			return
		}

		if len(comb)+(n-idx+1) < k {
			return
		}
		
		for i := idx; i <= n; i++ {
			comb = append(comb, i)
			backtrack(i+1, comb)
			comb = comb[:len(comb)-1]
		}
	}

	backtrack(1, []int{})
	return res
}
// @lc code=end

